### PR TITLE
Multi aave mai bal

### PR DESCRIPTION
**Pull request by viacheslavvbogdanov**

### CODE REVIEW

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        uint256[] memory toClaim = new uint256[](1); //TODO
+        return toClaim;
+    }
+
+    /// @dev Stub function for Strategy Base implementation
+    function poolTotalAmount() external view override returns (uint256) {
+        return _totalAmount;
+    }
+
+    function assets() external view override returns (address[] memory) {
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:140]: move it to strategy implementation

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        return _assets;
+    }
+
+    function platform() external pure override returns (Platform) {
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:144]: add AAVE_MAI_BAL to IStrategy

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    /// @dev Stub function for Strategy Base implementation
+    function withdrawAndClaimFromPool(uint256 underlyingAmount) internal override {
+        pumpOutSource(underlyingAmount, 0);
+        _totalAmount = calculator.getTotalAmountOut(0);
+    }
+
+    /// @dev Stub function for Strategy Base implementation
+    function emergencyWithdrawFromPool() internal override {
+        //noop
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:120]: implement please
if no way to emergency, do just withdraw

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    string public constant VERSION = "1.0.0";
+    /// @dev Placeholder, for non full buyback need to implement liquidation
+    uint256 private constant _BUY_BACK_RATIO = 10000;
+
+    /// @dev Assets should reflect underlying tokens for investing
+    address[] private _assets;
+
+    address public constant WMATIC = 0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270;
+
+    uint256 private _totalAmount = 0;
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:46]: add comment with explanation

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        segments.push(PipeSegment(maiStablecoinBorrowPipe, maiStablecoinBorrowPipe.create(maiStablecoinBorrowPipeData)));
+
+        BalVaultPipe balVaultPipe = new BalVaultPipe();
+        segments.push(PipeSegment(balVaultPipe, balVaultPipe.create(balVaultPipeData)));
+
+    }
+
+    /// @dev Stub function for Strategy Base implementation
+    function rewardPoolBalance() public override pure returns (uint256 bal) {
+        bal = 0;
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:90]: no way to be zero

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    /// @dev Stub function for Strategy Base implementation
+    function doHardWork() external onlyNotPausedInvesting override restricted {
+        // withdrawAndClaimFromPool(0);
+        //emergencyWithdrawFromPool();
+        liquidateReward();
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:97]: do deposit all avalable undelrying

File: contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        rebalanceAllPipes();
+        _totalAmount = calculator.getTotalAmountOut(0);
+    }
+
+    function _balance(address token) internal view returns (uint256) {
```
Comment (belbix) [inline contracts/base/strategies/multi/AaveMaiBalStrategyBase.sol:102]: do you need this function?

File: contracts/base/strategies/multi/pipes/WrappingPipe.sol [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    /// @dev decodes context
+    function context(bytes memory c) internal pure returns (address WETH) {
+      (WETH) = abi.decode(c, (address));
+    }
+
+    /// @dev function for investing, deposits, entering, borrowing
+    function _put(bytes memory c, uint256 amount)
+    override public returns (uint256 output) {
```
Comment (belbix) [inline contracts/base/strategies/multi/pipes/WrappingPipe.sol:23]: why this functions are public?

Comment (belbix) [reply contracts/base/strategies/multi/pipes/WrappingPipe.sol:23]: payable?

Comment (belbix) [reply contracts/base/strategies/multi/pipes/WrappingPipe.sol:23]: probably comment is wrong and it is fine to be public, please describe what this function do in comment

Comment (viacheslavvbogdanov) [reply contracts/base/strategies/multi/pipes/WrappingPipe.sol:23]: changed to onlyOwner (owner = strategy contract)

Comment (viacheslavvbogdanov) [reply contracts/base/strategies/multi/pipes/WrappingPipe.sol:23]: delegate calls removed from it, added onlyOwner

**Pull request closed without merging**